### PR TITLE
[TABLE MODEL] Support user-defined aggregate function

### PR DESCRIPTION
**Pull request by Cpaulyz**
## Description

https://timechor.feishu.cn/docx/U6MYdc2QtobOMkxxgDec6hI6npb

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/grouped/GroupedUserDefinedAggregateAccumulator.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    }
+  }
+
+  @Override
+  public void evaluateIntermediate(int groupId, ColumnBuilder columnBuilder) {
+    checkArgument(
+        columnBuilder instanceof BinaryColumnBuilder,
+        "intermediate input and output of UDAF should be BinaryColumn");
+    if (stateArray.get(groupId) == null) {
+      columnBuilder.writeBinary(new Binary(new byte[0]));
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/grouped/GroupedUserDefinedAggregateAccumulator.java:117]: when will this if happens? It seems that each group will always have a state? If so, better throw IllegalStateException here.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/utils/UDFDataTypeTransformer.java [hunk 2]
```diff
@@ -122,19 +126,54 @@ public static org.apache.tsfile.read.common.type.Type transformUDFDataTypeToRead
...
       case BLOB:
+        return BlobType.BLOB;
       case STRING:
+        return StringType.STRING;
+      default:
+        throw new IllegalArgumentException("Invalid input: " + type);
+    }
+  }
+
+  public static org.apache.tsfile.read.common.type.Type transformTSDataTypeToReadType(
```
Comment (JackieTien97) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/utils/UDFDataTypeTransformer.java:151]: delete this method, you can use TypeFactory.getType instead

Comment (JackieTien97) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/utils/UDFDataTypeTransformer.java:151]: also delete transformReadTypeToTSDataType in this class, it's not used, and we also have a replacement method called InternalTypeManager.getTSDataType

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/UserDefinedAggregateFunctionAccumulator.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  @Override
+  public boolean hasFinalResult() {
+    return false;
+  }
+
+  @Override
+  public void addStatistics(Statistics[] statistics) {
+    // UDAF not support calculate from statistics now
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/UserDefinedAggregateFunctionAccumulator.java:115]: done

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/grouped/GroupedUserDefinedAggregateAccumulator.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        new RecordIterator(
+            Arrays.asList(arguments), inputDataTypes, arguments[0].getPositionCount());
+    int index = 0;
+    while (iterator.hasNext()) {
+      int groupId = groupIds[index++];
+      State state = getOrCreateState(groupId);
+      if (state == null) {
+        state = aggregateFunction.createState();
+        stateArray.set(groupId, state);
+      }
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/grouped/GroupedUserDefinedAggregateAccumulator.java:88]: done

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/grouped/GroupedUserDefinedAggregateAccumulator.java (hunk 1 above)
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/aggregation/grouped/GroupedUserDefinedAggregateAccumulator.java:117]: done

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/utils/UDFDataTypeTransformer.java (hunk 2 above)
Comment (Cpaulyz) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/utils/UDFDataTypeTransformer.java:151]: done

**Pull request merged**